File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.kt
import java.util.*
import kotlin.collections.LinkedHashMap

class Solution {
    fun solution(begin: String, target: String, words: Array<String>):Int {

        var answer = 0

        val graph = LinkedHashMap<String,LinkedList<String>>()
        val q:Queue<String> = LinkedList()
        val visited = setOf<String>()


        graph[begin] = LinkedList()
        if(!words.contains(target)) {
            return 0
        }

        for(item in words) {
            graph[item] = LinkedList()
            if(canTranslate(begin,item)) {
                graph[begin]!!.add(item)
                graph[item]!!.add(begin)
            }

        }

        for(i in words.indices) {
            for (j in i until words.size) {
                if (i == j) {
                    continue
                }
                if(canTranslate(words[i],words[j])) {
                    graph[words[i]]!!.add(words[j])
                    graph[words[j]]!!.add(words[i])
                }
            }
        }

        q.add(begin)
        visited.plus(begin)

        var temp:String



        

        var qSize:Int
        
        while(q.isNotEmpty()) {
            qSize =q.size

            repeat(qSize) {
                temp = q.poll()
                if(temp=="-1") {
                    answer++
                }
                else{
                    if(temp == target) {
                        return answer +1
                    }


                    for(item in graph[temp]!!) {
                        if(!visited.contains(item)){
                            q.add(item)
                            visited.plus(item)
                        }
                    }
                }

            }
            q.add("-1")


        }



        return 0

    }
    fun canTranslate(a:String, b:String): Boolean{
        var count = 0
        for(i in a.indices) {
            if(a[i]!=b[i]) {
                count++
            }
        }

        return count == 1
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    answer = 0
    if target not in words:
        return 0
    
    graph = {}
    graph[begin] = []
    
    for item in words:
        graph[item]= []
        
    
    for i in range(len(words)):
        for j in range(i,len(words)):
            if check_can_go(words[i],words[j]):
                graph[words[i]].append(words[j])
                graph[words[j]].append(words[i])
    
    if begin not in words:
        for item in words:
            if check_can_go(begin,item):
                graph[item].append(begin)
                graph[begin].append(item)
    
    for key,value in graph.items():
        print(key, value)
        
    answer = search(graph,begin,target)
        
        
    
    return answer


def check_can_go(start,target):
    count = 0
    size = len(start)
    for i in range(size):
        if start[i] == target[i]:
            count+=1
    return count == size -1 

def search(graph,begin,target):
    s = set([begin])
    q = deque()
    q.append((begin,0))
    
    while q:
        node,count = q.popleft()
        for item in graph[node]:
            if node == target:
                return count
            if item in s:
                continue
            q.append((item,count+1))
            s.add(item)
    
    return 0
        